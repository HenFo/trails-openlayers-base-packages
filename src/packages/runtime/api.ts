// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ApiMethod = (...args: any[]) => any;

export interface ApiExtension {
    getApiMethods(): Promise<Record<string, ApiMethod>>;
}

export interface ApiService {
    getApi(): Promise<Record<string, ApiMethod>>;
}

declare module "./ServiceRegistry" {
    interface ServiceRegistry {
        "runtime.ApiExtension": ApiExtension;
        "runtime.ApiService": ApiService;
    }
}
