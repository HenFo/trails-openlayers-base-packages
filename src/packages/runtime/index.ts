export * from "./CustomElement";
export * as metadata from "./Metadata";
export * from "./Service";
