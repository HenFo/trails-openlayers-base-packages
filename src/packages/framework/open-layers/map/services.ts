// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { createLogger } from "@open-pioneer/core";
import { Service, ServiceOptions, ServiceType } from "@open-pioneer/runtime";
import TileLayer from "ol/layer/Tile";
import OlMap, { MapOptions } from "ol/Map";
import OSM from "ol/source/OSM";
import View from "ol/View";
import { OpenlayersMapConfigurationProvider } from "./api";

const defaultLayer = new TileLayer({ source: new OSM(), properties: { title: "OSM" } });

const LOG = createLogger("open-layers:OlMapRegistry");

interface References {
    providers: ServiceType<"open-layers-map-config.MapConfigProvider">[];
}

export class OlMapRegistry implements Service {
    private maps: Map<string, OlMap> = new Map();
    private configProvider: Map<string, OpenlayersMapConfigurationProvider> = new Map();

    constructor(options: ServiceOptions<References>) {
        const providers = options.references.providers;
        for (const provider of providers) {
            this.configProvider.set(provider.mapId, provider);
        }
    }

    destroy(): void {
        LOG.info(`Destroy map registry and all maps`);
        this.maps.forEach((map) => {
            map.dispose();
        });
        this.maps.clear();
    }

    async getMap(mapId: string): Promise<OlMap> {
        if (!this.maps.has(mapId)) {
            let additionalMapOptions = this.configProvider.get(mapId)?.getMapOptions();
            if (!additionalMapOptions) {
                LOG.warn(`config provider for map with id ${mapId} does not exist`);
                additionalMapOptions = {};
            }
            const defaultOptions: MapOptions = {
                layers: [defaultLayer],
                view: new View({
                    projection: "EPSG:3857",
                    center: [0, 0],
                    zoom: 1
                }),
                ...additionalMapOptions
            };
            LOG.info(`Create map with id '${mapId}'`);
            this.maps.set(mapId, new OlMap(defaultOptions));
        }
        const map = this.maps.get(mapId);
        if (!map) {
            throw new Error(`Map with id ${mapId} does not exist`);
        }
        return map;
    }

    setContainer(mapId: string, elem: HTMLDivElement) {
        const map = this.maps.get(mapId);
        if (map) {
            map.setTarget(elem);
            return;
        } else {
            throw new Error(`Map with id ${mapId} does not exist`);
        }
    }
}

declare module "@open-pioneer/runtime" {
    interface ServiceRegistry {
        "open-layers-map-service": OlMapRegistry;
    }
}
