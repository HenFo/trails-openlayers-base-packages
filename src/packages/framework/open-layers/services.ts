// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import Map from "ol/Map";
import { Service } from "@open-pioneer/runtime";
import { createManualPromise, ManualPromise } from "@open-pioneer/runtime/utils";

export class OpenLayersMapService implements Service {
    private map: Map | undefined = undefined;

    private mapPromise: ManualPromise<Map> | undefined;

    getMap(): Promise<Map> {
        if (this.map) {
            return Promise.resolve(this.map);
        }
        const mapPromise = (this.mapPromise ??= createManualPromise());
        return mapPromise.promise;
    }

    deleteMap(mapId: string) {
        console.log(`deleteMap ${mapId}`);
    }

    setMap(id: string, map: Map) {
        this.map = map;
        this.mapPromise?.resolve(this.map);
    }
}

declare module "@open-pioneer/runtime" {
    interface ServiceRegistry {
        "open-layers-map-service": OpenLayersMapService;
    }
}
