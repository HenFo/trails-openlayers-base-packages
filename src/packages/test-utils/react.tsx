import { Service } from "@open-pioneer/runtime";
import {
    PackageContext as InternalPackageContext,
    PackageContextMethods
} from "@open-pioneer/runtime/react-integration";
import { FC, ReactNode, useMemo } from "react";

export type AnyService = Service<Record<string, unknown>>;

export interface PackageContextProviderProps {
    /** Interface implementations, keyed by interface name. */
    services?: {
        [interfaceName: string]: AnyService;
    };

    /** Interface implementations, keyed by interface name and then by qualifier. */
    qualifiedServices?: {
        [interfaceName: string]: {
            [qualifier: string]: AnyService;
        };
    };

    /** Package properties (keyed by package name). */
    properties?: {
        [packageName: string]: Record<string, unknown>;
    };

    /** Children to render */
    children?: ReactNode;
}

/**
 * Injects services and properties into the component tree.
 * React components using pioneer hooks like `useService` and `useProperties`
 * will receive the mocked properties here instead.
 */
export const PackageContextProvider: FC<PackageContextProviderProps> = (props) => {
    const { services, qualifiedServices, properties, children } = props;
    const contextMethods = useMemo(
        () => createPackageContextMethods({ services, qualifiedServices, properties }),
        [services, qualifiedServices, properties]
    );
    return (
        <InternalPackageContext.Provider value={contextMethods}>
            {children}
        </InternalPackageContext.Provider>
    );
};

function createPackageContextMethods(
    options: Omit<PackageContextProviderProps, "children">
): PackageContextMethods {
    const services = options?.services ?? {};
    const qualifiedServices = options?.qualifiedServices ?? {};
    const properties = options?.properties ?? {};
    return {
        getService(packageName, interfaceName, options) {
            if (!options.qualifier) {
                const service = services[interfaceName];
                if (service) {
                    return service;
                }
                const qualified = qualifiedServices[interfaceName];
                if (qualified) {
                    for (const value of Object.values(qualified)) {
                        return value; // Return any implementation
                    }
                }
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}'. Update the configuration of PackageContextProvider.`
                );
            }

            const qualifiedService = qualifiedServices[interfaceName]?.[options.qualifier];
            if (!qualifiedService) {
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}' (qualifier '${options.qualifier}'). Update the configuration of PackageContextProvider.`
                );
            }
            return qualifiedService;
        },
        getServices(packageName, interfaceName) {
            const unqualified = services[interfaceName];
            const results = Object.values(qualifiedServices[interfaceName] ?? {});
            if (unqualified) {
                results.push(unqualified);
            }
            return results;
        },
        getProperties(packageName) {
            const packageProperties = properties[packageName];
            if (!packageProperties) {
                throw new Error(
                    `No properties for package '${packageName}' bound for testing. Update the configuration of PackageContextProvider.`
                );
            }
            return packageProperties;
        },
        getIntl(packageName) {
            // TODO
            throw new Error(`No i18n messages for package '${packageName}' bound for testing.`);
        }
    };
}
