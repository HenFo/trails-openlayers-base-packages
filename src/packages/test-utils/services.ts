import { Service, ServiceConstructor } from "@open-pioneer/runtime";

/**
 * Options for the {@link createService} function.
 */
export interface CreateServiceOptions<References> {
    /**
     * References provided to the service constructor.
     *
     * @default {}
     */
    references?: PartialServiceReferences<References>;

    /**
     * Properties provided to the service constructor.
     *
     * @default {}
     */
    properties?: Record<string, unknown>;
}

type PartialServiceReferences<References> = {
    [referenceName in keyof References]?: Partial<References[referenceName]>;
};

/**
 * Creates a new service instance of `clazz`.
 *
 * The options passed into the constructor of `clazz` are taking from the `options` parameter.
 *
 * This function allows the user to define mock- or test objects as service references.
 *
 * Most type-checking features are disabled for convenience, but by default this function's types
 * enforce that a value provided for a reference must match the partial interface of the actually required type.
 *
 * For example, if a service depends on `{ ctx: ApplicationContext }`, this function ensures
 * that the test value for `ctx` implements `Partial<ApplicationContext>`.
 * This can of course be disabled by casting the value explicitly.
 *
 * @param clazz a service class
 * @param options options passed to the service constructor
 * @returns a new instance of the given service class
 */
export async function createService<References extends {}, Interface extends {}>(
    clazz: ServiceConstructor<References, Interface>,
    options?: CreateServiceOptions<References>
): Promise<Service<Interface>> {
    return new clazz({
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        references: (options?.references ?? {}) as any,
        properties: options?.properties ?? {}
    });
}
