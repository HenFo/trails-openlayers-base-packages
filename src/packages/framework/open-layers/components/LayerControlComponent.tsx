// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import {
    Checkbox,
    Slider,
    SliderFilledTrack,
    SliderThumb,
    SliderTrack,
    Tooltip
} from "@open-pioneer/chakra-integration";
import Layer from "ol/layer/Layer";
import { useService } from "open-pioneer:react-hooks";
import { useEffect, useState } from "react";
import { useAsync } from "react-use";

function LayerVisibilityTogglerComponent(props: { layer: Layer }) {
    const [visibility, setVisibility] = useState<boolean>(props.layer.getVisible());

    const [title, setTitle] = useState<string>();

    useEffect(() => {
        setTitle(props.layer.getProperties().title ?? "undefined layer title");
    }, [props.layer]);

    const changeVisibility = () => {
        setVisibility(!visibility);
        props.layer.setVisible(!visibility);
    };

    return (
        <Checkbox size="lg" isChecked={visibility} onChange={changeVisibility}>
            {title}
        </Checkbox>
    );
}

function LayerOpacitySliderComponent(props: { layer: Layer }) {
    const [sliderValue, setSliderValue] = useState(props.layer.getOpacity() * 100);
    const [showTooltip, setShowTooltip] = useState(false);

    useEffect(() => {
        const curr = props.layer.getOpacity() * 100;
        if (curr !== sliderValue) {
            props.layer.setOpacity(sliderValue / 100);
        }
    }, [props.layer, sliderValue]);

    return (
        <Slider
            id="slider"
            defaultValue={sliderValue}
            min={0}
            max={100}
            colorScheme="teal"
            onChange={(v) => setSliderValue(v)}
            onMouseEnter={() => setShowTooltip(true)}
            onMouseLeave={() => setShowTooltip(false)}
        >
            <SliderTrack>
                <SliderFilledTrack />
            </SliderTrack>
            <Tooltip
                hasArrow
                bg="teal.500"
                color="white"
                placement="top"
                isOpen={showTooltip}
                label={`${sliderValue}%`}
            >
                <SliderThumb bg="teal.500"></SliderThumb>
            </Tooltip>
        </Slider>
    );
}

export function LayerControlComponent(config: { showopacitySlider: boolean }) {
    const mapPromise = useService("open-layers-map-service").getMap();

    const state = useAsync(async () => {
        const map = await mapPromise;
        const layers = map.getAllLayers();
        return layers;
    });

    return (
        <div>
            {state.loading ? (
                <div>Loading...</div>
            ) : state.error ? (
                <div>Error: {state.error.message}</div>
            ) : (
                <div>
                    {state.value?.map((layer, i) => (
                        <div key={i}>
                            <LayerVisibilityTogglerComponent
                                layer={layer}
                            ></LayerVisibilityTogglerComponent>
                            {config.showopacitySlider && (
                                <div>
                                    <LayerOpacitySliderComponent
                                        layer={layer}
                                    ></LayerOpacitySliderComponent>
                                </div>
                            )}
                        </div>
                    ))}
                </div>
            )}
        </div>
    );
}
