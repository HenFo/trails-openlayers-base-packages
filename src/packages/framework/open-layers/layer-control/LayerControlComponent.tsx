// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import {
    Checkbox,
    Slider,
    SliderFilledTrack,
    SliderThumb,
    SliderTrack,
    Tooltip
} from "@open-pioneer/chakra-integration";
import Layer from "ol/layer/Layer";
import { useIntl, useService } from "open-pioneer:react-hooks";
import { useEffect, useState } from "react";
import { useAsync } from "react-use";

import { OlComponentConfig } from "@open-pioneer/open-layers-map";

function LayerVisibilityTogglerComponent(props: { layer: Layer }) {
    const intl = useIntl();
    const [visibility, setVisibility] = useState<boolean>(props.layer.getVisible());
    const title =
        props.layer.getProperties().title ?? intl.formatMessage({ id: "undefined-layer-title" });

    const changeVisibility = () => {
        setVisibility(!visibility);
        props.layer.setVisible(!visibility);
    };

    return (
        <Checkbox size="lg" isChecked={visibility} onChange={changeVisibility}>
            {title}
        </Checkbox>
    );
}

function LayerOpacitySliderComponent(props: { layer: Layer }) {
    const [sliderValue, setSliderValue] = useState(props.layer.getOpacity() * 100);
    const [showTooltip, setShowTooltip] = useState(false);

    useEffect(() => {
        const curr = props.layer.getOpacity() * 100;
        if (curr !== sliderValue) {
            props.layer.setOpacity(sliderValue / 100);
        }
    }, [props.layer, sliderValue]);

    return (
        <Slider
            id="slider"
            defaultValue={sliderValue}
            min={0}
            max={100}
            colorScheme="teal"
            onChange={(v) => setSliderValue(v)}
            onMouseEnter={() => setShowTooltip(true)}
            onMouseLeave={() => setShowTooltip(false)}
        >
            <SliderTrack>
                <SliderFilledTrack />
            </SliderTrack>
            <Tooltip
                hasArrow
                bg="teal.500"
                color="white"
                placement="top"
                isOpen={showTooltip}
                label={`${sliderValue}%`}
            >
                <SliderThumb bg="teal.500"></SliderThumb>
            </Tooltip>
        </Slider>
    );
}

interface LayerControlConfig extends OlComponentConfig {
    showOpacitySlider: boolean;
}

export function LayerControlComponent(config: LayerControlConfig) {
    const olMapRegistry = useService("open-layers-map-service");

    const state = useAsync(async () => {
        const map = await olMapRegistry.getMap(config.mapId);
        const layers = map.getAllLayers().reverse();
        return layers;
    }, []);

    return (
        <div>
            {state.loading ? (
                <div>Loading...</div>
            ) : state.error ? (
                <div>Error: {state.error.message}</div>
            ) : (
                <div>
                    {state.value?.map((layer, i) => (
                        <div key={i}>
                            <LayerVisibilityTogglerComponent
                                layer={layer}
                            ></LayerVisibilityTogglerComponent>
                            {config.showOpacitySlider && (
                                <div>
                                    <LayerOpacitySliderComponent
                                        layer={layer}
                                    ></LayerOpacitySliderComponent>
                                </div>
                            )}
                        </div>
                    ))}
                </div>
            )}
        </div>
    );
}
