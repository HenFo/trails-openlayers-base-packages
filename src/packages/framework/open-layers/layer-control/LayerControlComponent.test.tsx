// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
/**
 * @vitest-environment jsdom
 */
import { OlMapConfigurationProvider } from "@open-pioneer/ol-map";
import { OlMapRegistry } from "@open-pioneer/ol-map/services";
import { ServiceOptions } from "@open-pioneer/runtime";
import {
    PackageContextProvider,
    PackageContextProviderProps
} from "@open-pioneer/test-utils/react";
import { createService } from "@open-pioneer/test-utils/services";
import { render, screen } from "@testing-library/react";
import TileLayer from "ol/layer/Tile";
import { MapOptions } from "ol/Map";
import Stamen from "ol/source/Stamen";
import { expect, it } from "vitest";
import { LayerControlComponent } from "./LayerControlComponent";

// used to avoid a "ResizeObserver is not defined" error
global.ResizeObserver = require("resize-observer-polyfill");

class MapConfigProvider implements OlMapConfigurationProvider {
    mapId = "default";
    mapOptions: MapOptions = {};

    constructor(options: ServiceOptions) {
        if (options.properties.mapOptions) {
            this.mapOptions = options.properties.mapOptions as MapOptions;
        }
        if (options.properties.mapId) {
            this.mapId = options.properties.mapId as string;
        }
    }

    getMapOptions(): Promise<MapOptions> {
        return Promise.resolve(this.mapOptions);
    }
}

it("should successfully create a layer control component", async () => {
    const mapConfigProvider = await createService(MapConfigProvider, {
        properties: {
            mapOptions: {
                layers: [
                    new TileLayer({
                        source: new Stamen({ layer: "watercolor" }),
                        properties: { title: "Watercolor" },
                        visible: false
                    })
                ]
            } as MapOptions,
            mapId: "test"
        }
    });

    const service = await createService(OlMapRegistry, {
        references: {
            providers: [mapConfigProvider]
        }
    });

    const mocks: PackageContextProviderProps = {
        services: {
            "ol-map.MapRegistry": service
        }
    };

    const renderResult = render(
        <PackageContextProvider {...mocks}>
            <div data-testid="base">
                <LayerControlComponent mapId="test" />
            </div>
        </PackageContextProvider>
    );

    // Assert layer control is mounted
    const div = await screen.findByTestId("base");
    expect(div).toMatchSnapshot();

    // Check if two layer controls for the configured layers are created
    const layerElems = renderResult.container.querySelectorAll(".layer-entry");
    expect(layerElems.length).toBe(1);
    expect(layerElems[0]).toBeInstanceOf(HTMLDivElement);
});
