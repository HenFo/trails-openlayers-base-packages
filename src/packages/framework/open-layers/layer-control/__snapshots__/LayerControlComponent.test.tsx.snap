// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should successfully create a layer control component 1`] = `
<div
  data-testid="base"
>
  <div>
    <div>
      <div
        class="layer-entry"
      >
        <label
          class="chakra-checkbox layer-select css-1elxtiq"
        >
          <input
            class="chakra-checkbox__input"
            style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
            type="checkbox"
            value=""
          />
          <span
            aria-hidden="true"
            class="chakra-checkbox__control css-dnty2r"
          />
          <span
            class="chakra-checkbox__label css-1oeb2oe"
          >
            Watercolor
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;
